### PR TITLE
fixed sign-compare warnings and some refactor

### PR DESCRIPTION
**Pull request by aldenml**
for preliminary review

### CODE REVIEW

File: include/libtorrent/sliding_average.hpp [hunk 1]
```diff
@@ -90,8 +92,9 @@ struct average_accumulator
 
 	void add_sample(int s)
 	{
+		TORRENT_ASSERT(s >= 0);
 		++m_num_samples;
-		m_sample_sum += s;
```
Comment (arvidn) [inline include/libtorrent/sliding_average.hpp]: I wonder if there's a good reason why this only allows positive samples. It seems a bit arbitrary. what would you think about making ``m_sample_sum`` signed instead?

File: src/file.cpp [hunk 2]
```diff
@@ -1364,13 +1364,13 @@ namespace libtorrent
...
 		{
-			std::uint32_t ret = 1;
-			for (int i = 0; i < str.size(); ++i)
+			std::int32_t ret = 1;
+			for (auto ch : str)
 			{
-				if (str[i] == 0) continue;
-					ret *= int(str[i]);
+				if (ch == 0) continue;
+				ret *= int(ch);
```
Comment (arvidn) [inline src/file.cpp]: I don't think this is valid. With ``ret`` being signed, it's undefined behavior to overflow it, which this multiplication most likely will do. instead, cast the character to ``std::uint32_t``

File: src/kademlia/get_item.cpp [hunk 3]
```diff
@@ -200,7 +200,7 @@ void get_item_observer::reply(msg const& m)
 	bdecode_node q = r.dict_find_int("seq");
 	if (q)
 	{
-		seq = sequence_number(q.int_value());
+		seq = sequence_number(std::uint64_t(q.int_value()));
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: would it make sense to make ``sequence_number`` be backed by a signed integer instead?

File: src/kademlia/node_id.cpp [hunk 4]
```diff
@@ -210,8 +211,8 @@ node_id generate_prefix_mask(int bits)
 	TORRENT_ASSERT(bits <= 160);
 	node_id mask(nullptr);
 	int b = 0;
```
Comment (arvidn) [inline src/kademlia/node_id.cpp]: if you make ``b`` unsigned instead, could you avoid the cast to ``size_t``?

File: src/kademlia/traversal_algorithm.cpp [hunk 5]
```diff
@@ -207,22 +207,23 @@ void traversal_algorithm::add_entry(node_id const& id
 
 	if (m_results.size() > 100)
 	{
-		for (int i = 100; i < int(m_results.size()); ++i)
+		std::for_each(m_results.begin() + 100, m_results.end()
+			, [this](std::shared_ptr<observer>& ptr)
```
Comment (arvidn) [inline src/kademlia/traversal_algorithm.cpp]: I think this could be a ``std::shared_ptr<observer> const& ptr``, right?

File: src/part_file.cpp [hunk 6]
```diff
@@ -118,8 +118,9 @@ namespace libtorrent
 
 			for (int i = 0; i < num_pieces; ++i)
 			{
-				int slot = read_uint32(ptr);
-				if (slot == 0xffffffff) continue;
+				std::uint32_t const uslot = read_uint32(ptr);
+				if (uslot == 0xffffffff) continue;
+				int const slot = int(uslot);
```
Comment (arvidn) [inline src/part_file.cpp]: I would imagine there's very little, if any at all, arithmetic on ``slot``. perhaps it could be an unsigned type (perhaps similar to ``allocation_slot`` in stack_allocator). making it a special type should probably be a separate patch though

File: src/piece_picker.cpp [hunk 7]
```diff
@@ -413,7 +413,7 @@ namespace libtorrent
 		for (int i = 0; i < num_pieces; ++i)
 		{
 			int h = have[i];
```
Comment (arvidn) [inline src/piece_picker.cpp]: would it make more sense to widen ``h`` here instead of truncating the result from ``count()``?

File: src/piece_picker.cpp [hunk 8]
```diff
@@ -496,7 +496,7 @@ namespace libtorrent
 				block_info const* info = blocks_for_piece(dp);
 				for (int k = 0; k < num_blocks; ++k)
 				{
-					TORRENT_ASSERT(info[k].piece_index == dp.index);
+					TORRENT_ASSERT(info[k].piece_index == int(dp.index));
```
Comment (arvidn) [inline src/piece_picker.cpp]: hm.. why are these incompatible? is ``dp.index`` a bitfield?

**Comment by arvidn** [comment]
looks reasonable for the most part. do you understand why the tests for the openssl builds are failing?

File: include/libtorrent/sliding_average.hpp (hunk 1 above)
Comment (aldenml) [inline include/libtorrent/sliding_average.hpp]: I think it makes sense

File: src/file.cpp (hunk 2 above)
Comment (aldenml) [inline src/file.cpp]: true, I will do it

File: src/kademlia/node_id.cpp (hunk 4 above)
Comment (aldenml) [inline src/kademlia/node_id.cpp]: it could be, I will try it.

File: src/piece_picker.cpp (hunk 7 above)
Comment (aldenml) [inline src/piece_picker.cpp]: to review

**Comment by aldenml** [comment]
no idea why `gcc+openssl` is failing, more likely I introduced a subtle bug, I will make the comments changes and rebase, but I will hunt the crash in my travis account

**Comment by arvidn** [comment]
sorry about the conflicts...

**Comment by aldenml** [comment]
it's OK :)

File: include/libtorrent/session_handle.hpp [hunk 9]
```diff
@@ -407,7 +407,7 @@ namespace libtorrent
 		// 	To generate the signature, you may want to use the
 		// 	``sign_mutable_item`` function.
 		//
-		// std::uint64_t& seq
```
Comment (arvidn) [inline include/libtorrent/session_handle.hpp]: this is an API change. it should probably be mentioned in the Changelog
